### ISSUE TITLE
How To Run Google Datalab Notebooks

### ISSUE THREAD

**Issue by xifengwang** [GoogleCloudPlatform/datalab#657]
Hello!
    I have registered on Google Developers Console, but my project is not a billed project. I did the steps of  “initialized environment.”  and “Build and Run
”as the web pages https://github.com/GoogleCloudPlatform/datalab/wiki/Development-Environment and https://github.com/GoogleCloudPlatform/datalab/wiki/Build-and-Run describe.
    When I run code 
"SELECT * FROM [cloud-datalab-samples:httplogs.logs_20140615] LIMIT 3" 
in Google Cloud SDK bq shell (installed locally) ,I can get the correct result:
mydatalab-1108> query 'SELECT TOP(title, 10) as title, COUNT(*) as revisions FROM [publicdata:samples.wikipedia] WHERE wp_namespace = 0'
Waiting on bqjob_r258aa7e960555512_0000015099c21972_1 ... (0s) Current status: DONE    
+-------------------------------------------------+-----------+
|                      title                      | revisions |
+-------------------------------------------------+-----------+
| George W. Bush                                  |     43652 |
| List of World Wrestling Entertainment employees |     30551 |
| Wikipedia                                       |     29702 |
| United States                                   |     27411 |
| Michael Jackson                                 |     23178 |
| Jesus                                           |     21707 |
| Deaths in 2009                                  |     20679 |
| World War II                                    |     20449 |
| Britney Spears                                  |     20356 |
| Wii                                             |     20100 |
+-------------------------------------------------+-----------+
    But when I run it in notebook (deployed locally), I get the following error:
Unable to load metadata from metadata service.
    I wonder whether I forget configure something for Notebooks in Google Developers Console,or I do not login notebook when I connect to http://myip:8081/notebooks/datalab/tutorials/BigQuery/Hello%20BigQuery.ipynb from mybrowser(because i do not know how to login from this page).
When I type "gcloud auth print-access-token" in linux shell,  I can get my token printed.
    It seems that I run into the same question as https://github.com/GoogleCloudPlatform/datalab/issues/651
    Please give me some ideas! Thank you! I am anxious to Google Cloud Datalab.

**Comment by xifengwang** [comment]
I have registered on Google Developers Console, and my project is a billed project. I did the steps of “initialized environment.” and “Build and Run ”as the web pages https://github.com/GoogleCloudPlatform/datalab/wiki/Development-Environment and https://github.com/GoogleCloudPlatform/datalab/wiki/Build-and-Run describe. But when i run code in Notebook deployed on my local linux server,i run into the following error:

bq.Query('SELECT * FROM [cloud-datalab-samples:httplogs.logs_20140615] LIMIT 3').results()

Exception Traceback (most recent call last) in () 1 # Create and run a SQL query ----> 2 bq.Query('SELECT * FROM [cloud-datalab-samples:httplogs.logs_20140615] LIMIT 3').results()

/usr/local/lib/python2.7/dist-packages/gcp/bigquery/_query.pyc in results(self, use_cache) 130 """ 131 if not use_cache or (self._results is None): --> 132 self.execute(use_cache=use_cache) 133 return self._results.results 134

/usr/local/lib/python2.7/dist-packages/gcp/bigquery/_query.pyc in execute(self, table_name, table_mode, use_cache, priority, allow_large_results) 343 """ 344 job = self.execute_async(table_name=table_name, table_mode=table_mode, use_cache=use_cache, --> 345 priority=priority, allow_large_results=allow_large_results) 346 self._results = job.wait() 347 return self._results

/usr/local/lib/python2.7/dist-packages/gcp/bigquery/_query.pyc in execute_async(self, table_name, table_mode, use_cache, priority, allow_large_results) 307 allow_large_results=allow_large_results) 308 except Exception as e: --> 309 raise e 310 if 'jobReference' not in query_result: 311 raise Exception('Unexpected query response.')

Exception: Failed to send HTTP request. 

 Step by step,I find the place which throws the exception: 
if headers is None: headers = {}

headers['user-agent'] = 'GoogleCloudDataLab/1.0'
# Add querystring to the URL if there are any arguments.
if args is not None:
  qs = urllib.urlencode(args)
  url = url + '?' + qs

# Setup method to POST if unspecified, and appropriate request headers
# if there is data to be sent within the request.
if data is not None:
  if method is None:
    method = 'POST'

  if data != '':
    # If there is a content type specified, use it (and the data) as-is.
    # Otherwise, assume JSON, and serialize the data object.
    if 'Content-Type' not in headers:
      data = json.dumps(data)
      headers['Content-Type'] = 'application/json'
  headers['Content-Length'] = str(len(data))
else:
  if method == 'POST':
    headers['Content-Length'] = '0'

# If the method is still unset, i.e. it was unspecified, and there
# was no data to be POSTed, then default to GET request.
if method is None:
  method = 'GET'

# Create an Http object to issue requests. Associate the credentials
# with it if specified to perform authorization.
#
# TODO(nikhilko):
# SSL cert validation seemingly fails, and workarounds are not amenable
# to implementing in library code. So configure the Http object to skip
# doing so, in the interim.
http = httplib2.Http()
http.disable_ssl_certificate_validation = True
if credentials is not None:
http = credentials.authorize(http)

try:
  response, content = http.request(url,method=method,body=data,headers=headers)
  if 200 <= response.status < 300:
    if raw_response:
      return content
    return json.loads(content)
  else:
    raise RequestException(response.status, content)
except ValueError:
  raise Exception('Failed to process HTTP response.')
except httplib2.HttpLib2Error:
  raise Exception('Failed to send HTTP request.')

  It seems that I can not send http request in my docker container,but I can not find the reason.I wonder whether it is my configuration error or The cloud datalab does not support deploy&run in my local docker container(but the github repository shows the datalab can run locally).That is to say,we cannot run code in notebooks on local datalab server. 
As far as I can say,if the developer want to use resources on Google Cloud Platform such as VMs,he must pay for the VMs.But if he installs the google cloud sdk(I can find bq,gsutil,etc. in $GCLOID_SDK_INSTALL_DIR/bin) , but he uses his own VM(docker container),the datalab can also run&deploy.Am I right?(^_^).
    Please give me some ideas.The question has disturbed me for one week!Thank you!

 Best Regards!

**Comment by xifengwang** [comment]
Hello Javier !
    Thank you for your reply! 
    I have registered on Google Developers Console, and my project is a billed project. I did the steps of “initialized environment.” and “Build and Run ”as the web pages https://github.com/GoogleCloudPlatform/datalab/wiki/Development-Environment and https://github.com/GoogleCloudPlatform/datalab/wiki/Build-and-Run describe. But when i run code in Notebook deployed on my local linux server,i run into the following error:

bq.Query('SELECT * FROM [cloud-datalab-samples:httplogs.logs_20140615] LIMIT 3').results()

Exception Traceback (most recent call last) in () 1 # Create and run a SQL query ----> 2 bq.Query('SELECT * FROM [cloud-datalab-samples:httplogs.logs_20140615] LIMIT 3').results()

/usr/local/lib/python2.7/dist-packages/gcp/bigquery/_query.pyc in results(self, use_cache) 130 """ 131 if not use_cache or (self._results is None): --> 132 self.execute(use_cache=use_cache) 133 return self._results.results 134

/usr/local/lib/python2.7/dist-packages/gcp/bigquery/_query.pyc in execute(self, table_name, table_mode, use_cache, priority, allow_large_results) 343 """ 344 job = self.execute_async(table_name=table_name, table_mode=table_mode, use_cache=use_cache, --> 345 priority=priority, allow_large_results=allow_large_results) 346 self._results = job.wait() 347 return self._results

/usr/local/lib/python2.7/dist-packages/gcp/bigquery/_query.pyc in execute_async(self, table_name, table_mode, use_cache, priority, allow_large_results) 307 allow_large_results=allow_large_results) 308 except Exception as e: --> 309 raise e 310 if 'jobReference' not in query_result: 311 raise Exception('Unexpected query response.')

Exception: Failed to send HTTP request. 

 Step by step,I find the place which throws the exception: 
if headers is None: headers = {}

headers['user-agent'] = 'GoogleCloudDataLab/1.0'
# Add querystring to the URL if there are any arguments.
if args is not None:
  qs = urllib.urlencode(args)
  url = url + '?' + qs

# Setup method to POST if unspecified, and appropriate request headers
# if there is data to be sent within the request.
if data is not None:
  if method is None:
    method = 'POST'

  if data != '':
    # If there is a content type specified, use it (and the data) as-is.
    # Otherwise, assume JSON, and serialize the data object.
    if 'Content-Type' not in headers:
      data = json.dumps(data)
      headers['Content-Type'] = 'application/json'
  headers['Content-Length'] = str(len(data))
else:
  if method == 'POST':
    headers['Content-Length'] = '0'

# If the method is still unset, i.e. it was unspecified, and there
# was no data to be POSTed, then default to GET request.
if method is None:
  method = 'GET'

# Create an Http object to issue requests. Associate the credentials
# with it if specified to perform authorization.
#
# TODO(nikhilko):
# SSL cert validation seemingly fails, and workarounds are not amenable
# to implementing in library code. So configure the Http object to skip
# doing so, in the interim.
http = httplib2.Http()
http.disable_ssl_certificate_validation = True
if credentials is not None:
http = credentials.authorize(http)

try:
  response, content = http.request(url,method=method,body=data,headers=headers)
  if 200 <= response.status < 300:
    if raw_response:
      return content
    return json.loads(content)
  else:
    raise RequestException(response.status, content)
except ValueError:
  raise Exception('Failed to process HTTP response.')
except httplib2.HttpLib2Error:
  raise Exception('Failed to send HTTP request.')

  It seems that I can not send http request in my docker container,but I can not find the reason.I wonder whether it is my configuration error or The cloud datalab does not support deploy&run in my local docker container(but the github repository shows the datalab can run locally).That is to say,we cannot run code in notebooks on local datalab server. 
As far as I can say,if the developer want to use resources on Google Cloud Platform such as VMs,he must pay for the VMs.But if he installs the google cloud sdk(I can find bq,gsutil,etc. in $GCLOID_SDK_INSTALL_DIR/bin) , but he uses his own VM(docker container),the datalab can also run&deploy.Am I right?(^_^).
    Please give me some ideas.The question has disturbed me for one week!Thank you!

 Best Regards!

**Comment by gramster** [comment]
This seems like a duplicate of #651
